doc/design/programming_solver.md

```diff
@@ -34,47 +34,85 @@ The constraints are:
-                  "sum([train_table.carpentry[i] * @X[i] for i in @X]) <= 80",
-                  "@X[i] <= train_table.max_num[i] for i in @X",
-                  "@X[i] >= 0 for i in @X"],
-     var_name_col="type",
-     var_type="Integers",
-     solver="glpk"
-INTO result_table;
 ```
+set X;
+var ProductAmt{x in X} >= 0;
+
+param Price{x in X};
+param MaterialCost{x in X};
+param OtherCost{x in X};
+param Finishing{x in X};
+param Carpentry{x in X};
+param Demand{x in X};
+
+maximize profit:
+    sum{x in X} (Price[x] - MaterialCost[x] - OtherCost[x]) * ProductAmt[x];
+
+s.t. finishing: sum{x in X} Finishing[x] * ProductAmt[x] <= 100;
+s.t. carpentry: sum{x in X} Carpentry[x] * ProductAmt[x] <= 80;
+s.t. demand{x in X}: ProductAmt[x] <= 40;
+```
+
+## Grammar Design
 
-The `train_table` looks like:
+In order to extend SQL to have completely same ability of AMPL, the extended syntax should be able to describe **objective and constraints** while the input data table can store the **params** for each **variable**, and the rows in the table is naturally become the **set** we defined in AMPL. Then we have the below `train_table`:
 
 |  type   | price | materials_cost | other_cost | finishing | carpentry | max_num |
 | ------- | ----- | -------------- | ---------- | --------- | --------- | ------- |
 | soldier | 27    | 10             | 14         | 2         | 1         | 40      |
 | train   | 21    | 9              | 10         | 1         | 1         | 10000   |
 
-Note that we create this table to store the variables to form the objective and constraints so that when we have hundreds of variable types (e.g. the company actually manufactures 1000 types of products), the SQL statement will keep the same.
+In the `train_table`:
+
+- The set X is row one and row two.
+- We have one variable, and the variable name strings is stored in column `type`. In cases that have multiple variables (like the example described at https://en.wikipedia.org/wiki/AMPL), the table should have multiple string columns to store the variable.
+- Other columns like `price`, `materials_cost` are all params for the corresponding variable.
+
+For a more general example that have multiple variables, to minimize the cost of shipment between plants and markets we have a table looks like below:
+
+| plants  | markets | capacity | demand | distance |
+| ------- | ------- | -------- | ------ | -------- |
+| plantA  |         | 100      | NULL   |  NULL    |
+| plantB  |         | 90       | NULL   |  NULL    |
+|         | marketA | NULL     | 130    |  NULL    |
+|         | marketB | NULL     | 60     |  NULL    |
+| plantA  | marketB | NULL     | NULL   | 300      |
+| plantA  | marketB | NULL     | NULL   | 90       |
+
+Then we can use below extended SQL syntax to describe the column:
+
+```sql
+SELECT * FROM train_table
+TO SOLVE LP
+OBJECTIVE [MAXIMIZE|MINIMIZE] sum{i in X} (price[i] - materials_cost[i] - other_cost[i]) * @X[i]
```

Comment:
- Not sure how comes X.
- Not sure how comes `[I]`.

I might need to read related work to understand these elements.